### PR TITLE
Add support for SDMXXX energy meters

### PR DESCRIPTION
**Pull request by polyfaces**
## Description:

Add support for SDM220M energy meter

**Related issue (if applicable):** fixes

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#745

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by dcec** [comment]
Have sense to use all SDM library https://github.com/reaper7/SDM_Energy_Meter?

**Comment by polyfaces** [comment]
> 
> 
> Have sense to use all SDM library https://github.com/reaper7/SDM_Energy_Meter?

I am opened to use the SDM library.  However, in my point of view, I think it is better to reuse existing modbus components in esphome.

**Comment by dcec** [comment]
I found this request: https://github.com/esphome/feature-requests/issues/289.
have sense to add the support of all SDM family and start with the more popolar  SDM120/SDM630.
I'm available to test with SDM120 if you want

File: esphome/components/sdm220m/sdm220m.cpp [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+  temp = sdm220m_get_32bit(148);
+  float export_active_energy = *((float *) &temp);
+
+  temp = sdm220m_get_32bit(152);
+  float import_reactive_energy = *((float *) &temp);
+
+  temp = sdm220m_get_32bit(156);
+  float export_reactive_energy = *((float *) &temp);
+
+  ESP_LOGD(TAG,
```
Comment (jesserockz) [inline esphome/components/sdm220m/sdm220m.cpp:63]: This probably does not need to be a debug message.

```suggestion
  ESP_LOGV(TAG,
```

File: esphome/components/sdm220m/sensor.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    CONF_CURRENT, CONF_EXPORT_ACTIVE_ENERGY, \
+    CONF_EXPORT_REACTIVE_ENERGY, CONF_FREQUENCY, CONF_ID, \
+    CONF_IMPORT_ACTIVE_ENERGY, CONF_IMPORT_REACTIVE_ENERGY, \
+    CONF_PHASE_ANGLE, CONF_POWER_FACTOR, CONF_REACTIVE_POWER, \
+    CONF_VOLTAGE, ICON_COUNTER, ICON_CURRENT_AC, ICON_FLASH, \
+    ICON_POWER, UNIT_AMPERE, UNIT_DEGREES, UNIT_EMPTY, UNIT_HERTZ, \
+    UNIT_VOLT, UNIT_VOLT_AMPS, UNIT_VOLT_AMPS_REACTIVE, \
+    UNIT_VOLT_AMPS_REACTIVE_HOURS, UNIT_WATT, UNIT_WATT_HOURS
+
+AUTO_LOAD = ['modbus']
```
Comment (jesserockz) [inline esphome/components/sdm220m/sensor.py:14]: Please add yourself as a codeowner and then run `script/build_codeowners.py`

```suggestion
AUTO_LOAD = ['modbus']
CODEOWNERS = ['@polyfaces']
```

**Review by jesserockz — DISMISSED** [review]
Thanks, only 2 things below then it's ready to merge.

**Comment by reaper7** [comment]
why do you limit addon (at least in terms of nomenclature) to 220m ?

I've prepared a list of registries in a table form:
https://github.com/reaper7/SDM_Energy_Meter/blob/master/SDM.h#L95

Maybe you will choose common registers and prepare a more general version?

**Comment by dcec** [comment]
@reaper7  is also possible to use rs232<->rs485 converter with additional pins for flow control, like MAX485 with esphome?

**Comment by reaper7** [comment]
@dcec - why not, You can try with custom component: https://github.com/reaper7/esphome_sdm_custom

Remember to specify which pin will be responsible for flow control:

uncomment this line and replace NOT_A_PIN with the correct pin number
https://github.com/reaper7/esphome_sdm_custom/blob/master/custom_components/sdm/SDM_Config_User.h#L52

and replace NOT_A_PIN with the correct pin number(the same as above) in this line:
https://github.com/reaper7/esphome_sdm_custom/blob/master/custom_components/sdm/sdm_esphome.h#L5

**Comment by corvis** [comment]
+1 for making this more generic and adding more registers/sensors at least in the future. E.g. I have SDM220 and SDM630 (3 phases). Both are currently connected directly to the home assistant via USB<>ModBus converer but it would be geart to connect them to the esphome.

@polyfaces are there any wiring instructions how to connect to esphome? It's modbus, so I some hardware converter is required. Maybe it makes sence to cover this in docs?

**Comment by nikito7** [comment]
https://github.com/esphome/esphome/pull/1571

**Comment by jesserockz** [comment]
@polyfaces I hope you don't mind, I made a few changes to this component...

`sdm220m` is now `sdm_meter` and it supports sensors for the 3 phases which allows support for the sdm6XX series.

@reaper7 @corvis if you guys could test this out, that would be great. I have the SDM230M here working fantastically with this.

I also added a `flow_control_pin` to the `sdm_meter` as I personally have a `max485` connected.

Cheers, Jesse

**Review by jesserockz — APPROVED** [review]

**Comment by reaper7** [comment]
work ok for me (sdm220) but I see one error:
```
[19:41:32][E][uart:054]:   Invalid stop bits: Integration requested stop_bits 2 but you have 1!
```
my uart config:
```
uart:
  rx_pin: GPIO3
  tx_pin: GPIO1
  baud_rate: 9600
  stop_bits: 1
```

sdm meter is also set to 1 stop bit...

I remember that somewhere in the esphome-> modbus code 2 stop bits were permanently set

**Comment by jesserockz** [comment]
Good point, I think I will remove that and leave it up to the individual component to verify the settings.

**Pull request merged**